### DISCUSSION TITLE
TypeId Feature Discussion

### DISCUSSION THREAD

**Discussion by drusellers** [SteveDunn/Vogen#615] [Ideas]
I was reading [Designing APIs for humans: Object IDs](https://dev.to/stripe/designing-apis-for-humans-object-ids-3o5a) looking at [TypeId](https://github.com/firenero/TypeId) and wondered if there was any appetite for this library to add a concept like this. 

Specifically this idea of the `type_` prefix to whatever the underlying type is.

```csharp
[ValueObject<Guid>(Prefix = "cust")]
public partial struct CustomerId {
}
```

I've really been enjoying using the idea of `CustomerId` over `Guid` in my code base, and the idea of going back to even something as nice as `TypeId` feels like a step backwards from Vogen. 

So before I go down a path of either writing it myself, or looking at extending vogen I thought I would just ask and see what y'all thought first. 

Cheers.

**Comment by SteveDunn** [comment]
Thanks for the link to that article, it's an interesting read. I certainly think this would be useful in Vogen. I initially thought Vogen would just be useful in the domain layer, but it's also very popular in the infrastructure layer for things like IDs.

The `prefix` suggested would mean that the underlying type would no longer be a (.NET) GUID, much like the examples given in the article also aren't .NET GUIDs.

So, we'd want something like this, but trimmed down (controlled :

```c#
[ValueObject<string>]
public partial struct CustomerId
{
    private static Validation Validate(string input) => 
        input.StartsWith("cus_") ? Validation.Ok : Validation.Invalid("Customer IDs start with cus_");

    public static CustomerId NewUnique() => From($"cus_{Guid.NewGuid()}");
}
```

... or using TypeId...

```c#
[ValueObject<TypeIdDecoded>]
public partial struct CustomerId
{
    private static Validation Validate(TypeIdDecoded input) => 
        input.Type.StartsWith("cus") ? Validation.Ok : Validation.Invalid("Customer IDs start with cus_");

    public static CustomerId NewUnique() => From(TypeId.New("cus"));
}
```

Assuming this solution was adequate, we might still have issues with serialization.

I'll continue to think about this and find an elegant solution.

**Reply by drusellers** [reply]
> The prefix suggested would mean that the underlying type would no longer be a (.NET) GUID, much like the examples given in the article also aren't .NET GUIDs.

Agreed, I would end up storing them in the database as text fields (PG)

I've thought of an alternate where in the database they are still persisted as uuid, and the prefix is stripped off. But that might be more effort than its worth, and might be confusing to others when they don't see the expected uuid.

```csharp
[ValueObject<string>(Prefix = "cus")]
public partial struct CustomerId
{
    // source generated would be preferred
    private static Validation Validate(string input) => 
        input.StartsWith("cus_") ? Validation.Ok : Validation.Invalid("Customer IDs start with cus_");

    // source generated preferred, but yes.
    public static CustomerId NewUnique() => From($"cus_{Guid.NewGuid()}");
}
```

Additionally, I'm also interested in, but not a hard requirement at all, in variable length random numbers rather than the longer Guid. So that I could have something like `cus_deadbeef` where I know the over all number of items is low and I want the id to be used in a URL

**Reply by drusellers** [reply]
@SteveDunn just wanted to follow back up on this. Does this still feel like a valid idea for Vogen?

**Reply by SteveDunn** [reply]
Hi @drusellers - seeing as Vogen seems to be used quite a lot for IDs, I do still think it's a valid idea for Vogen to support this.

Random thoughts:

* If a 'prefix' parameter is used, it essentially turns the underlying type into a string
* Unless a user-supplied `Validate` method is provided, Vogen will generate one to ensure that the input string has the correct prefix
* Vogen will generate a `NewUnique` method consisting of a string that starts with the prefix, following by some sort of random/unique characters

It's the last bullet point which needs clarification. We'd need some sort of extensibility point to delegate to something that generates these. It's a shame that attributes can't take expressions. Perhaps we could use interceptors?

**Reply by drusellers** [reply]
1. Agreed - 100% - I almost want to say that if you set the prefix and the type "T" isn't string it should blow up.
2. Love that
3. Love that.

```csharp
[ValueObject<string>(Prefix = "cus_")]
public partial struct CustomerId
{
    // User Can Override
    // source generated would be preferred
    private static Validation Validate(string input) => 
        input.StartsWith("cus_") ? Validation.Ok : Validation.Invalid($"{nameOf(CustomerId)} start with {Prefix}");


    // 1. the NewUnique Method
    // source generated preferred, but yes.
    public static CustomerId NewUnique() => From($"{Prefix}{GeneratePostfix()}");
    // which then has the validate to confirm

    // 2. or a function that just generates the random bits and Vogen source generates the combined "NewMethod"
    // if not provided...
    // default source generators for Guid = Guid.NewGuid()
    // 
    // for string I might do
    private static string GeneratePostfix() => RandomHashFunction.Next();
}
```

Oh, what about...
```csharp
// where RandomString is a known interface?
[ValueObject<string>(Prefix = "cus_", Generator = typeof(RandomString)]
public partial struct CustomerId;
```

**Reply by SteveDunn** [reply]
It doesn't look too difficult in this light. I'll try to get around to it, but it probably won't be for a few weeks. Thanks again for the feedback and idea - I think it'll be useful to a lot of people.

**Reply by drusellers** [reply]
Hello again 👋 - just checking in on this.

**Reply by SteveDunn** [reply]
Apologies @drusellers - I'm flat out with 'real work' (unfortunately). I'll hope to return to Vogen in a  few weeks when things quiten down.

**Reply by drusellers** [reply]
No worries! Just checking in. 🙂🙂
